File: robot_model/scripts/cam_test_csi.py
import cv2
import numpy as np

# Left camera pipeline (CAM0)
pipeline_left = (
    "nvarguscamerasrc sensor-id=0 ! "
    "video/x-raw(memory:NVMM), width=640, height=480, framerate=30/1, format=NV12 ! "
    "nvvidconv ! video/x-raw, format=BGRx ! "
    "videoconvert ! video/x-raw, format=BGR ! appsink"
)

# Right camera pipeline (CAM1)
pipeline_right = (
    "nvarguscamerasrc sensor-id=1 ! "
    "video/x-raw(memory:NVMM), width=640, height=480, framerate=30/1, format=NV12 ! "
    "nvvidconv ! video/x-raw, format=BGRx ! "
    "videoconvert ! video/x-raw, format=BGR ! appsink"
)

cap_left = cv2.VideoCapture(pipeline_left, cv2.CAP_GSTREAMER)
cap_right = cv2.VideoCapture(pipeline_right, cv2.CAP_GSTREAMER)

if not cap_left.isOpened() or not cap_right.isOpened():
    print("Could not open one or both cameras!")
    exit()

while True:
    retL, frameL = cap_left.read()
    retR, frameR = cap_right.read()

    if not retL or not retR:
        print("Frame read failed!")
        break

    # 🔥 Flip both frames vertically
    frameL = cv2.flip(frameL, 0)   # 0 → vertical flip
    frameR = cv2.flip(frameR, 0)

    # Resize both frames
    frameL = cv2.resize(frameL, (320, 240))
    frameR = cv2.resize(frameR, (320, 240))

    # Combine side-by-side
    combined = np.hstack((frameL, frameR))

    cv2.imshow("Stereo Cameras (Flipped)", combined)

    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cap_left.release()
cap_right.release()
cv2.destroyAllWindows()

File: scripts/cam_test_csi.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <cstdlib>   // for system()
#include <thread>
#include <chrono>

int main()
{
    std::cout << "🔄 Restarting nvargus-daemon..." << std::endl;

    int ret = system("sudo systemctl restart nvargus-daemon");

    if (ret != 0)
    {
        std::cerr << "❌ Failed to restart nvargus-daemon" << std::endl;
        return -1;
    }

    // ⏳ Give Argus time to come up
    std::this_thread::sleep_for(std::chrono::seconds(2));

    // ================= CAMERA PIPELINES =================

    std::string pipeline_left =
        "nvarguscamerasrc sensor-id=0 maxperf-enable=1 ! "
        "video/x-raw(memory:NVMM), width=640, height=480, framerate=30/1, format=NV12 ! "
        "nvvidconv flip-method=0 ! "
        "video/x-raw, format=BGRx ! "
        "videoconvert ! video/x-raw, format=BGR ! appsink drop=1";

    std::string pipeline_right =
        "nvarguscamerasrc sensor-id=1 maxperf-enable=1 ! "
        "video/x-raw(memory:NVMM), width=640, height=480, framerate=30/1, format=NV12 ! "
        "nvvidconv flip-method=0 ! "
        "video/x-raw, format=BGRx ! "
        "videoconvert ! video/x-raw, format=BGR ! appsink drop=1";

    cv::VideoCapture cap_left(pipeline_left, cv::CAP_GSTREAMER);

    // 🔥 CRITICAL: delay before opening second camera
    std::this_thread::sleep_for(std::chrono::milliseconds(800));

    cv::VideoCapture cap_right(pipeline_right, cv::CAP_GSTREAMER);

    if (!cap_left.isOpened() || !cap_right.isOpened())
    {
        std::cerr << "❌ Could not open one or both cameras!" << std::endl;
        return -1;
    }

    cv::Mat frameL, frameR, combined;

    while (true)
    {
        if (!cap_left.read(frameL) || !cap_right.read(frameR))
        {
            std::cerr << "❌ Frame read failed!" << std::endl;
            break;
        }

        cv::flip(frameL, frameL, 0);
        cv::flip(frameR, frameR, 0);

        cv::resize(frameL, frameL, cv::Size(320, 240));
        cv::resize(frameR, frameR, cv::Size(320, 240));

        cv::hconcat(frameL, frameR, combined);
        cv::imshow("Stereo Cameras", combined);

        if (cv::waitKey(1) == 'q')
            break;
    }

    cap_left.release();
    cap_right.release();
    cv::destroyAllWindows();

    return 0;
}
